server/v2/stf/stf.go

```diff
@@ -108,17 +104,48 @@ func (s STF[T]) validateTx(ctx context.Context, state store.WritableState, gasLi
 	return validateCtx.gasUsed, validateCtx.events, applyStateChanges(state, validateState)
 }
 
+// execTx executes the tx messages on the provided state. If the tx fails then the state is discarded.
 func (s STF[T]) execTx(ctx context.Context, state store.WritableState, gasLimit uint64, tx T) (msgResp Type, gasUsed uint64, execEvents []event.Event, err error) {
 	execState := s.branch(state)
 	execCtx := s.makeContext(ctx, tx.GetSenders(), execState, gasLimit)
 	// atomic execution of the all messages in a transaction, TODO: we should allow messages to fail in a specific mode
-	for _, msg := range tx.GetMessages() {
-		msgResp, err = s.handleMsg(execCtx, msg)
+	var txErr error
+	for i, msg := range tx.GetMessages() {
+		msgResp, txErr = s.handleMsg(execCtx, msg)
 		if err != nil {
```

Comment:
is this supposed to be `txErr`

Reply:
Actually, I find this method a bit hard to grok with how the errors are treated.

Reply:
100% good catch, we do not have any multi msg test flow (yet), as we're still understanding what's going to happen with multi msg, multi signers etc.